distribution/push.go

```diff
@@ -127,6 +133,19 @@ func Push(ctx context.Context, ref reference.Named, imagePushConfig *ImagePushCo
 			continue
 		}
 
+		parsedURL, urlErr := url.Parse(endpoint.URL)
```

Comment:
> Where should the URL parsing happen for this?

I really don't know why `APIEndpoint` stores `URL` as `string` and `Endpoint` stores `URL` as `*url.URL` but that looks like a smell tbh. Why don't we have a single place where all the parsing happens? Would that make all this logic easier?